File: api/Enkl.Api/Controllers/ChatController.cs
using Enkl.Api.Auth;
using Enkl.Api.Dtos;
using Enkl.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Enkl.Api.Controllers;

/// <summary>
/// Org-wide chat — [Authorize] only at class level (no ProjectMember/OrgAdmin policy), since every
/// org user can create channels/DMs and post messages; only the Truncate action below is further
/// restricted per-method to OrgAdmin. Mirrors TaskCommentsController's hand-spelled-routes shape
/// (nested under a specific resource, not a generic per-project entity helper) since this is also a
/// cross-cutting, non-generic pattern.
/// </summary>
[ApiController]
[Authorize]
[Route("api/chat")]
public class ChatController : ControllerBase
{
    private readonly ChatService _chat;
    private readonly SseBroadcaster _broadcaster;

    public ChatController(ChatService chat, SseBroadcaster broadcaster)
    {
        _chat = chat;
        _broadcaster = broadcaster;
    }

    private bool CallerIsOrgAdmin => User.HasClaim("orgAdmin", "true");
    private string CallerDisplayName => User.FindFirst("displayName")?.Value ?? "Someone";

    [HttpGet("org-users")]
    public async Task<IActionResult> GetOrgRoster()
    {
        return Ok(await _chat.GetOrgRosterAsync(User.OrgId()));
    }

    [HttpGet("channels")]
    public async Task<IActionResult> ListChannels()
    {
        return Ok(await _chat.ListChannelsAsync(User.OrgId(), User.UserId(), CallerIsOrgAdmin));
    }

    [HttpPost("channels")]
    public async Task<IActionResult> CreateChannel(CreateChatChannelRequest request)
    {
        var result = await _chat.CreateChannelAsync(User.OrgId(), User.UserId(), CallerDisplayName, request);
        return Ok(result);
    }

    [HttpPost("channels/{channelId:guid}/members")]
    public async Task<IActionResult> AddMember(Guid channelId, AddChatChannelMemberRequest request)
    {
        var ok = await _chat.AddMemberAsync(User.OrgId(), User.UserId(), CallerIsOrgAdmin, channelId, request.UserId);
        return ok ? NoContent() : NotFound();
    }

    [HttpDelete("channels/{channelId:guid}/members/{userId:guid}")]
    public async Task<IActionResult> RemoveMember(Guid channelId, Guid userId)
    {
        var ok = await _chat.RemoveMemberAsync(User.OrgId(), User.UserId(), CallerIsOrgAdmin, channelId, userId);
        return ok ? NoContent() : NotFound();
    }

    // Caller's own membership row only — 404 doubles as "not a member" and "channel doesn't exist",
    // same no-enumeration-oracle rule every other cross-tenant-ish check in this app follows.
    [HttpPut("channels/{channelId:guid}/mute")]
    public async Task<IActionResult> SetMuted(Guid channelId, SetChatChannelMuteRequest request)
    {
        var ok = await _chat.SetChannelMutedAsync(User.OrgId(), User.UserId(), channelId, request.IsMuted);
        return ok ? NoContent() : NotFound();
    }

    [HttpGet("search")]
    public async Task<IActionResult> Search([FromQuery] string q, [FromQuery] int limit = 20)
    {
        return Ok(await _chat.SearchAsync(User.OrgId(), User.UserId(), CallerIsOrgAdmin, q, limit));
    }

    [HttpGet("channels/{channelId:guid}/messages")]
    public async Task<IActionResult> GetMessages(Guid channelId, [FromQuery] DateTime? before, [FromQuery] int limit = 50)
    {
        var result = await _chat.GetMessagesAsync(User.OrgId(), User.UserId(), CallerIsOrgAdmin, channelId, before, limit);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPost("channels/{channelId:guid}/messages")]
    public async Task<IActionResult> PostMessage(Guid channelId, PostChatMessageRequest request)
    {
        var result = await _chat.PostMessageAsync(User.OrgId(), User.UserId(), CallerDisplayName, channelId, request);
        if (result is null) return NotFound();
        await BroadcastAsync(channelId, result.Value.Message, result.Value.ChannelMemberUserIds, "created");
        return Ok(result.Value.Message);
    }

    [HttpPut("channels/{channelId:guid}/messages/{messageId:guid}")]
    public async Task<IActionResult> UpdateMessage(Guid channelId, Guid messageId, UpdateChatMessageRequest request)
    {
        var result = await _chat.UpdateMessageAsync(User.UserId(), channelId, messageId, request);
        if (result is null) return NotFound();
        await BroadcastAsync(channelId, result.Value.Message, result.Value.ChannelMemberUserIds, "updated");
        return Ok(result.Value.Message);
    }

    [HttpDelete("channels/{channelId:guid}/messages/{messageId:guid}")]
    public async Task<IActionResult> DeleteMessage(Guid channelId, Guid messageId)
    {
        var result = await _chat.DeleteMessageAsync(User.OrgId(), User.UserId(), CallerIsOrgAdmin, channelId, messageId);
        if (result is null) return NotFound();
        await BroadcastAsync(channelId, result.Value.Message, result.Value.ChannelMemberUserIds, "deleted");
        return Ok(result.Value.Message);
    }

    [HttpPost("channels/{channelId:guid}/messages/{messageId:guid}/reactions")]
    public async Task<IActionResult> ToggleReaction(Guid channelId, Guid messageId, ToggleChatReactionRequest request)
    {
        var result = await _chat.ToggleReactionAsync(User.OrgId(), User.UserId(), CallerIsOrgAdmin, channelId, messageId, request.Emoji);
        if (result is null) return NotFound();
        await BroadcastReactionAsync(channelId, result.Value.Message, result.Value.ChannelMemberUserIds);
        return Ok(result.Value.Message);
    }

    // Org-Admin-only manual replacement for a scheduled 180-day purge (see ChatService.TruncateOldMessagesAsync's
    // own doc comment) — hard-deletes, no confirmation beyond the frontend's own confirm dialog.
    [HttpPost("truncate")]
    [Authorize(Policy = "OrgAdmin")]
    public async Task<IActionResult> Truncate()
    {
        return Ok(await _chat.TruncateOldMessagesAsync(User.OrgId()));
    }

    // Best-effort — a notification failure must never fail the mutation itself (same convention as
    // TasksController.BroadcastAsync).
    private async Task BroadcastAsync(Guid channelId, ChatMessageDto message, List<Guid> channelMemberUserIds, string changeType)
    {
        try
        {
            var clientSessionId = Request.Headers["X-Client-Session-Id"].FirstOrDefault();
            _broadcaster.BroadcastChatMessage(
                channelMemberUserIds,
                new ChatMessageEventDto(channelId, message.Id, message.Text, changeType, message.AuthorUserId, message.AuthorName, message.DateCreated, message.IsDeleted, message.MentionedUserIds),
                clientSessionId);
        }
        catch
        {
            // best-effort, see comment above
        }
    }

    private async Task BroadcastReactionAsync(Guid channelId, ChatMessageDto message, List<Guid> channelMemberUserIds)
    {
        try
        {
            var clientSessionId = Request.Headers["X-Client-Session-Id"].FirstOrDefault();
            _broadcaster.BroadcastChatReaction(
                channelMemberUserIds,
                new ChatReactionEventDto(channelId, message.Id, message.Reactions),
                clientSessionId);
        }
        catch
        {
            // best-effort, see comment above
        }
    }
}

File: php-api/src/Controllers/ChatController.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Controllers;

use Enkl\Api\Db\Database;
use Enkl\Api\Realtime\Broadcaster;
use Enkl\Api\Services\ChatService;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

/** Ported from Controllers/ChatController.cs. Org-wide chat — RequireAuthMiddleware only on the base
 * group (no ProjectMember/OrgAdmin), since every org user can create channels/DMs and post messages;
 * only the Truncate route (see routes.php) is further restricted to OrgAdmin. */
final class ChatController extends BaseController
{
    private function service(): ChatService
    {
        return new ChatService(Database::connection());
    }

    private function callerIsOrgAdmin(Request $request): bool
    {
        $claims = $request->getAttribute('jwtClaims');
        return ($claims->orgAdmin ?? null) === 'true';
    }

    public function getOrgRoster(Request $request, Response $response, array $args): Response
    {
        return $this->json($response, $this->service()->getOrgRoster($this->callerOrgId($request)));
    }

    public function listChannels(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->listChannels($this->callerOrgId($request), $this->callerUserId($request), $this->callerIsOrgAdmin($request));
        return $this->json($response, $result);
    }

    public function createChannel(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->createChannel(
            $this->callerOrgId($request), $this->callerUserId($request), $this->callerDisplayName($request) ?? 'Someone', $this->body($request)
        );
        return $this->json($response, $result);
    }

    public function addMember(Request $request, Response $response, array $args): Response
    {
        $body = $this->body($request);
        $ok = $this->service()->addMember(
            $this->callerOrgId($request), $this->callerUserId($request), $this->callerIsOrgAdmin($request), $args['channelId'], (string) ($body['userId'] ?? '')
        );
        return $ok ? $this->noContent($response) : $this->notFound($response);
    }

    public function removeMember(Request $request, Response $response, array $args): Response
    {
        $ok = $this->service()->removeMember(
            $this->callerOrgId($request), $this->callerUserId($request), $this->callerIsOrgAdmin($request), $args['channelId'], $args['userId']
        );
        return $ok ? $this->noContent($response) : $this->notFound($response);
    }

    // Caller's own membership row only — 404 doubles as "not a member" and "channel doesn't exist",
    // same no-enumeration-oracle rule every other cross-tenant-ish check in this app follows.
    public function setMuted(Request $request, Response $response, array $args): Response
    {
        $body = $this->body($request);
        $ok = $this->service()->setChannelMuted(
            $this->callerOrgId($request), $this->callerUserId($request), $args['channelId'], (bool) ($body['isMuted'] ?? false)
        );
        return $ok ? $this->noContent($response) : $this->notFound($response);
    }

    public function search(Request $request, Response $response, array $args): Response
    {
        $query = $request->getQueryParams();
        $term = (string) ($query['q'] ?? '');
        $limit = isset($query['limit']) ? (int) $query['limit'] : 20;
        $result = $this->service()->search($this->callerOrgId($request), $this->callerUserId($request), $this->callerIsOrgAdmin($request), $term, $limit);
        return $this->json($response, $result);
    }

    public function getMessages(Request $request, Response $response, array $args): Response
    {
        $query = $request->getQueryParams();
        $before = isset($query['before']) && $query['before'] !== '' ? (string) $query['before'] : null;
        $limit = isset($query['limit']) ? (int) $query['limit'] : 50;

        $result = $this->service()->getMessages(
            $this->callerOrgId($request), $this->callerUserId($request), $this->callerIsOrgAdmin($request), $args['channelId'], $before, $limit
        );
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function postMessage(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->postMessage(
            $this->callerOrgId($request), $this->callerUserId($request), $this->callerDisplayName($request) ?? 'Someone', $args['channelId'], $this->body($request)
        );
        if ($result === null) {
            return $this->notFound($response);
        }
        $this->broadcast($request, $args['channelId'], $result['message'], $result['channelMemberUserIds'], 'created');
        return $this->json($response, $result['message']);
    }

    public function updateMessage(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->updateMessage($this->callerUserId($request), $args['channelId'], $args['messageId'], $this->body($request));
        if ($result === null) {
            return $this->notFound($response);
        }
        $this->broadcast($request, $args['channelId'], $result['message'], $result['channelMemberUserIds'], 'updated');
        return $this->json($response, $result['message']);
    }

    public function deleteMessage(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->deleteMessage(
            $this->callerOrgId($request), $this->callerUserId($request), $this->callerIsOrgAdmin($request), $args['channelId'], $args['messageId']
        );
        if ($result === null) {
            return $this->notFound($response);
        }
        $this->broadcast($request, $args['channelId'], $result['message'], $result['channelMemberUserIds'], 'deleted');
        return $this->json($response, $result['message']);
    }

    public function toggleReaction(Request $request, Response $response, array $args): Response
    {
        $body = $this->body($request);
        $result = $this->service()->toggleReaction(
            $this->callerOrgId($request), $this->callerUserId($request), $this->callerIsOrgAdmin($request),
            $args['channelId'], $args['messageId'], (string) ($body['emoji'] ?? '')
        );
        if ($result === null) {
            return $this->notFound($response);
        }
        $this->broadcastReaction($request, $args['channelId'], $result['message'], $result['channelMemberUserIds']);
        return $this->json($response, $result['message']);
    }

    // Best-effort — a notification failure must never fail the mutation itself (same convention as
    // broadcast() above).
    private function broadcastReaction(Request $request, string $channelId, array $message, array $channelMemberUserIds): void
    {
        try {
            $clientSessionId = $request->getHeaderLine('X-Client-Session-Id') ?: null;
            (new Broadcaster(Database::connection()))->broadcastChatReaction(
                $channelMemberUserIds, $channelId, $message['id'], $message['reactions'], $clientSessionId
            );
        } catch (\Throwable) {
            // best-effort, see comment above
        }
    }

    // Org-Admin-only manual replacement for a scheduled 180-day purge (see ChatService::truncateOldMessages's
    // own doc comment) — hard-deletes, gated by OrgAdminMiddleware in routes.php.
    public function truncate(Request $request, Response $response, array $args): Response
    {
        return $this->json($response, $this->service()->truncateOldMessages($this->callerOrgId($request)));
    }

    // Best-effort — a notification failure must never fail the mutation itself (same convention as
    // TasksController.php's own broadcast call site).
    private function broadcast(Request $request, string $channelId, array $message, array $channelMemberUserIds, string $changeType): void
    {
        try {
            $clientSessionId = $request->getHeaderLine('X-Client-Session-Id') ?: null;
            (new Broadcaster(Database::connection()))->broadcastChatMessage(
                $channelMemberUserIds, $channelId, $message['id'], $message['text'], $changeType,
                $message['authorUserId'], $message['authorName'], $message['dateCreated'], $message['isDeleted'],
                $message['mentionedUserIds'], $clientSessionId
            );
        } catch (\Throwable) {
            // best-effort, see comment above
        }
    }
}
